scripts/process/as_svg.R

```diff
@@ -0,0 +1,266 @@
+#' and any other attributes added. See details
+
+#' @details 
+#' `<use>` assumes the there are attributes which point to re-use of elements 
+#' elsewhere in the svg
+#' if included, and if the `data` are spatial the `clip_box` removes features outside
+#' of the box. 
+#' 
+#' Additional attributes can be added by using the optional `attributes` field 
+#' in the viz block. These must be character key value pairs, but can use mustache 
+#' text substitution for simple modifications (see `svg_sub_attributes`).
+#' 
+process.as_svg_use <- function(viz){
+  
+  attributes <- viz[['attributes']]
+  
+  svg_use_out <- as_svg_elements('use', viz)$elements %>% 
+    svg_sub_attributes(attributes = attributes) %>% 
+    data.frame() %>% 
+    select_(.dots = c('.value', names(attributes)))
+  
+  saveRDS(svg_use_out, file = viz[['location']])
+}
+
+#' shared operations that occur for each process.as_svg_{element}
+#' 
+as_svg_elements <- function(element_name, viz){
+  deps <- readDepends(viz)
+  box <- deps[['clip_box']]
+  
+  elements <- clip_sf(deps[['data']], box) %>% 
+    mutate(.value = element_name) %>% 
+    select(.value, everything())
+  
+  append(list(elements = elements), plot_bounds(box))
+}
+
+#' takes output from one or many process.as_svg_{element} calls 
+#' and creates a `defs` svg container for it
+#' 
+#' @return writes an .rds file that contains the xml from `<g>` in a list
+#' format that can be converted to xml with xml2::as_xml_document()
+#' @details 
+#' if `depends` is empty, this still creates an element, it just 
+#' doesn't have xml children
+#' named `depends` are converted into `g` (group) names within `defs`
+process.as_svg_defs <- function(viz){
+  deps <- readDepends(viz)
+  defs <- xml_new_root('.x') 
+  group <- xml_add_child(defs, 'defs')
+  
+  for (g_name in names(deps)){
+    g <- xml_add_child(group, 'g', id = g_name)
+    svg_data <- deps[[g_name]]
+    for (j in seq_len(nrow(svg_data))){
+      do.call(xml_add_child, append(list(.x = g), svg_data[j, ]))
+    }
+  }
+  saveRDS(xml2::as_list(defs), file = viz[['location']])
+}
+
+#' create an xml/svg "group" or `<g>` element for each named dependency
+#' 
+#' @return writes an .rds file that contains the xml from `<g>` in a list
+#' format that can be converted to xml with xml2::as_xml_document()
+#' @details doesn't not name the id in the returned group 
+process.as_svg_g <- function(viz){
+  deps <- readDepends(viz)
+  g_main <- xml_new_root('.x') 
+  group <- do.call(xml_add_child, append(list(.x = g_main, .value = 'g'), viz[['attributes']]))
+  
+  for (g_name in names(deps)){
+    g <- xml_add_child(group, 'g', id = g_name)
+    svg_data <- deps[[g_name]]
+    for (j in 1:nrow(svg_data)){
+      do.call(xml_add_child, append(list(.x = g), svg_data[j, ]))
+    }
+  }
+
+  saveRDS(xml2::as_list(g_main), file = viz[['location']])
+}
+
+#' set up the basic plot par for a map
+set_svg_plot <- function(xlim, ylim){
```

Comment:
if it's setting up a basic plot par for a map, should the function name be more specific, e.g. `set_svg_map`?

Reply:
xlim/ylim are typically going to be in map units and the plot is creating a map typically. But the more I think about it, this isn't necessarily specific to plots that are maps. Maybe just take "map" out of the comment?